File: SkiKids/Models/DINTable.swift
import Foundation

enum DINCode: String, CaseIterable {
    case A, B, C, D, E, F, G, H, I, J, K, L, M

    var weightRange: ClosedRange<Int> {
        switch self {
        case .A: return 10...13
        case .B: return 14...17
        case .C: return 18...21
        case .D: return 22...25
        case .E: return 26...30
        case .F: return 31...35
        case .G: return 36...41
        case .H: return 42...48
        case .I: return 49...57
        case .J: return 58...66
        case .K: return 67...78
        case .L: return 79...94
        case .M: return 95...200
        }
    }

    var heightThreshold: Int {
        switch self {
        case .A: return 148
        case .B: return 148
        case .C: return 148
        case .D: return 148
        case .E: return 148
        case .F: return 157
        case .G: return 166
        case .H: return 174
        case .I: return 182
        case .J: return 189
        case .K: return 196
        case .L: return 196
        case .M: return Int.max
        }
    }

    var next: DINCode? {
        let all = DINCode.allCases
        guard let idx = all.firstIndex(of: self), idx + 1 < all.count else { return nil }
        return all[idx + 1]
    }

    var previous: DINCode? {
        let all = DINCode.allCases
        guard let idx = all.firstIndex(of: self), idx > 0 else { return nil }
        return all[idx - 1]
    }
}

enum BSLRange: Equatable {
    case upTo250
    case from251to270
    case from271to290
    case from291to310
    case from311to330
    case over330

    static func from(_ bsl: Int) -> BSLRange {
        if bsl <= 250 { return .upTo250 }
        if bsl <= 270 { return .from251to270 }
        if bsl <= 290 { return .from271to290 }
        if bsl <= 310 { return .from291to310 }
        if bsl <= 330 { return .from311to330 }
        return .over330
    }
}

struct DINTableEntry {
    let code: DINCode
    let bslRange: BSLRange
    let typeI: Double
    let typeII: Double
    let typeIII: Double
}

struct DINTable {
    static let entries: [DINTableEntry] = [
        // Code A (10-13kg)
        DINTableEntry(code: .A, bslRange: .upTo250,      typeI: 0.75, typeII: 0.75, typeIII: 1.00),
        DINTableEntry(code: .A, bslRange: .from251to270, typeI: 0.75, typeII: 0.75, typeIII: 1.00),
        DINTableEntry(code: .A, bslRange: .from271to290, typeI: 0.75, typeII: 1.00, typeIII: 1.25),
        DINTableEntry(code: .A, bslRange: .from291to310, typeI: 0.75, typeII: 1.00, typeIII: 1.25),
        DINTableEntry(code: .A, bslRange: .from311to330, typeI: 0.75, typeII: 1.00, typeIII: 1.50),
        DINTableEntry(code: .A, bslRange: .over330,      typeI: 0.75, typeII: 1.00, typeIII: 1.50),

        // Code B (14-17kg)
        DINTableEntry(code: .B, bslRange: .upTo250,      typeI: 1.00, typeII: 1.25, typeIII: 1.50),
        DINTableEntry(code: .B, bslRange: .from251to270, typeI: 1.00, typeII: 1.25, typeIII: 1.50),
        DINTableEntry(code: .B, bslRange: .from271to290, typeI: 1.00, typeII: 1.50, typeIII: 1.75),
        DINTableEntry(code: .B, bslRange: .from291to310, typeI: 1.00, typeII: 1.50, typeIII: 1.75),
        DINTableEntry(code: .B, bslRange: .from311to330, typeI: 1.25, typeII: 1.50, typeIII: 2.00),
        DINTableEntry(code: .B, bslRange: .over330,      typeI: 1.25, typeII: 1.50, typeIII: 2.00),

        // Code C (18-21kg)
        DINTableEntry(code: .C, bslRange: .upTo250,      typeI: 1.50, typeII: 1.75, typeIII: 2.25),
        DINTableEntry(code: .C, bslRange: .from251to270, typeI: 1.50, typeII: 1.75, typeIII: 2.25),
        DINTableEntry(code: .C, bslRange: .from271to290, typeI: 1.50, typeII: 2.00, typeIII: 2.50),
        DINTableEntry(code: .C, bslRange: .from291to310, typeI: 1.50, typeII: 2.00, typeIII: 2.50),
        DINTableEntry(code: .C, bslRange: .from311to330, typeI: 1.75, typeII: 2.25, typeIII: 3.00),
        DINTableEntry(code: .C, bslRange: .over330,      typeI: 1.75, typeII: 2.25, typeIII: 3.00),

        // Code D (22-25kg)
        DINTableEntry(code: .D, bslRange: .upTo250,      typeI: 2.00, typeII: 2.50, typeIII: 3.00),
        DINTableEntry(code: .D, bslRange: .from251to270, typeI: 2.00, typeII: 2.50, typeIII: 3.00),
        DINTableEntry(code: .D, bslRange: .from271to290, typeI: 2.00, typeII: 2.75, typeIII: 3.50),
        DINTableEntry(code: .D, bslRange: .from291to310, typeI: 2.00, typeII: 2.75, typeIII: 3.50),
        DINTableEntry(code: .D, bslRange: .from311to330, typeI: 2.25, typeII: 3.00, typeIII: 3.50),
        DINTableEntry(code: .D, bslRange: .over330,      typeI: 2.25, typeII: 3.00, typeIII: 3.50),

        // Code E (26-30kg)
        DINTableEntry(code: .E, bslRange: .upTo250,      typeI: 2.50, typeII: 3.00, typeIII: 3.50),
        DINTableEntry(code: .E, bslRange: .from251to270, typeI: 2.50, typeII: 3.00, typeIII: 3.50),
        DINTableEntry(code: .E, bslRange: .from271to290, typeI: 2.75, typeII: 3.50, typeIII: 4.00),
        DINTableEntry(code: .E, bslRange: .from291to310, typeI: 2.75, typeII: 3.50, typeIII: 4.00),
        DINTableEntry(code: .E, bslRange: .from311to330, typeI: 3.00, typeII: 3.50, typeIII: 4.50),
        DINTableEntry(code: .E, bslRange: .over330,      typeI: 3.00, typeII: 3.50, typeIII: 4.50),

        // Code F (31-35kg)
        DINTableEntry(code: .F, bslRange: .upTo250,      typeI: 3.00, typeII: 3.50, typeIII: 4.50),
        DINTableEntry(code: .F, bslRange: .from251to270, typeI: 3.00, typeII: 3.50, typeIII: 4.50),
        DINTableEntry(code: .F, bslRange: .from271to290, typeI: 3.50, typeII: 4.00, typeIII: 5.00),
        DINTableEntry(code: .F, bslRange: .from291to310, typeI: 3.50, typeII: 4.00, typeIII: 5.00),
        DINTableEntry(code: .F, bslRange: .from311to330, typeI: 3.50, typeII: 4.50, typeIII: 5.50),
        DINTableEntry(code: .F, bslRange: .over330,      typeI: 3.50, typeII: 4.50, typeIII: 5.50),

        // Code G (36-41kg)
        DINTableEntry(code: .G, bslRange: .upTo250,      typeI: 3.50, typeII: 4.50, typeIII: 5.50),
        DINTableEntry(code: .G, bslRange: .from251to270, typeI: 3.50, typeII: 4.50, typeIII: 5.50),
        DINTableEntry(code: .G, bslRange: .from271to290, typeI: 4.00, typeII: 5.00, typeIII: 6.00),
        DINTableEntry(code: .G, bslRange: .from291to310, typeI: 4.00, typeII: 5.00, typeIII: 6.00),
        DINTableEntry(code: .G, bslRange: .from311to330, typeI: 4.50, typeII: 5.50, typeIII: 6.50),
        DINTableEntry(code: .G, bslRange: .over330,      typeI: 4.50, typeII: 5.50, typeIII: 6.50),

        // Code H (42-48kg)
        DINTableEntry(code: .H, bslRange: .upTo250,      typeI: 4.50, typeII: 5.50, typeIII: 6.50),
        DINTableEntry(code: .H, bslRange: .from251to270, typeI: 4.50, typeII: 5.50, typeIII: 6.50),
        DINTableEntry(code: .H, bslRange: .from271to290, typeI: 5.00, typeII: 6.00, typeIII: 7.50),
        DINTableEntry(code: .H, bslRange: .from291to310, typeI: 5.00, typeII: 6.00, typeIII: 7.50),
        DINTableEntry(code: .H, bslRange: .from311to330, typeI: 5.50, typeII: 6.50, typeIII: 8.00),
        DINTableEntry(code: .H, bslRange: .over330,      typeI: 5.50, typeII: 6.50, typeIII: 8.00),

        // Code I (49-57kg)
        DINTableEntry(code: .I, bslRange: .upTo250,      typeI: 5.50, typeII: 6.50, typeIII: 8.00),
        DINTableEntry(code: .I, bslRange: .from251to270, typeI: 5.50, typeII: 6.50, typeIII: 8.00),
        DINTableEntry(code: .I, bslRange: .from271to290, typeI: 6.00, typeII: 7.50, typeIII: 9.00),
        DINTableEntry(code: .I, bslRange: .from291to310, typeI: 6.00, typeII: 7.50, typeIII: 9.00),
        DINTableEntry(code: .I, bslRange: .from311to330, typeI: 6.50, typeII: 8.00, typeIII: 9.50),
        DINTableEntry(code: .I, bslRange: .over330,      typeI: 6.50, typeII: 8.00, typeIII: 9.50),

        // Code J (58-66kg)
        DINTableEntry(code: .J, bslRange: .upTo250,      typeI: 6.50, typeII: 8.00, typeIII: 9.50),
        DINTableEntry(code: .J, bslRange: .from251to270, typeI: 6.50, typeII: 8.00, typeIII: 9.50),
        DINTableEntry(code: .J, bslRange: .from271to290, typeI: 7.50, typeII: 9.00, typeIII: 10.50),
        DINTableEntry(code: .J, bslRange: .from291to310, typeI: 7.50, typeII: 9.00, typeIII: 10.50),
        DINTableEntry(code: .J, bslRange: .from311to330, typeI: 8.00, typeII: 9.50, typeIII: 11.00),
        DINTableEntry(code: .J, bslRange: .over330,      typeI: 8.00, typeII: 9.50, typeIII: 11.00),

        // Code K (67-78kg)
        DINTableEntry(code: .K, bslRange: .upTo250,      typeI: 8.00,  typeII: 9.50,  typeIII: 11.00),
        DINTableEntry(code: .K, bslRange: .from251to270, typeI: 8.00,  typeII: 9.50,  typeIII: 11.00),
        DINTableEntry(code: .K, bslRange: .from271to290, typeI: 8.50,  typeII: 10.00, typeIII: 12.00),
        DINTableEntry(code: .K, bslRange: .from291to310, typeI: 8.50,  typeII: 10.00, typeIII: 12.00),
        DINTableEntry(code: .K, bslRange: .from311to330, typeI: 9.00,  typeII: 11.00, typeIII: 12.00),
        DINTableEntry(code: .K, bslRange: .over330,      typeI: 9.00,  typeII: 11.00, typeIII: 12.00),

        // Code L (79-94kg)
        DINTableEntry(code: .L, bslRange: .upTo250,      typeI: 9.00,  typeII: 11.00, typeIII: 12.00),
        DINTableEntry(code: .L, bslRange: .from251to270, typeI: 9.00,  typeII: 11.00, typeIII: 12.00),
        DINTableEntry(code: .L, bslRange: .from271to290, typeI: 9.50,  typeII: 11.00, typeIII: 12.00),
        DINTableEntry(code: .L, bslRange: .from291to310, typeI: 9.50,  typeII: 11.00, typeIII: 12.00),
        DINTableEntry(code: .L, bslRange: .from311to330, typeI: 10.00, typeII: 12.00, typeIII: 12.00),
        DINTableEntry(code: .L, bslRange: .over330,      typeI: 10.00, typeII: 12.00, typeIII: 12.00),

        // Code M (95+kg)
        DINTableEntry(code: .M, bslRange: .upTo250,      typeI: 10.00, typeII: 12.00, typeIII: 12.00),
        DINTableEntry(code: .M, bslRange: .from251to270, typeI: 10.00, typeII: 12.00, typeIII: 12.00),
        DINTableEntry(code: .M, bslRange: .from271to290, typeI: 11.00, typeII: 12.00, typeIII: 12.00),
        DINTableEntry(code: .M, bslRange: .from291to310, typeI: 11.00, typeII: 12.00, typeIII: 12.00),
        DINTableEntry(code: .M, bslRange: .from311to330, typeI: 12.00, typeII: 12.00, typeIII: 12.00),
        DINTableEntry(code: .M, bslRange: .over330,      typeI: 12.00, typeII: 12.00, typeIII: 12.00),
    ]

    static func lookup(code: DINCode, bsl: Int, ability: AbilityLevel) -> Double {
        let bslRange = BSLRange.from(bsl)
        guard let entry = entries.first(where: { $0.code == code && $0.bslRange == bslRange }) else {
            return 1.0
        }
        switch ability {
        case .beginner: return entry.typeI
        case .intermediate: return entry.typeII
        case .advanced: return entry.typeIII
        }
    }

}

File: react-native/src/calculators/din-table.ts
import { AbilityLevel, BSLRange, DINCode, DINTableEntry } from "../models/types";

export const DIN_CODE_ORDER: DINCode[] = [
  DINCode.A,
  DINCode.B,
  DINCode.C,
  DINCode.D,
  DINCode.E,
  DINCode.F,
  DINCode.G,
  DINCode.H,
  DINCode.I,
  DINCode.J,
  DINCode.K,
  DINCode.L,
  DINCode.M,
];

export interface DINCodeMetadata {
  weightMin: number;
  weightMax: number;
  heightThreshold: number;
}

export const DIN_CODE_METADATA: Record<DINCode, DINCodeMetadata> = {
  [DINCode.A]: { weightMin: 10, weightMax: 13, heightThreshold: 148 },
  [DINCode.B]: { weightMin: 14, weightMax: 17, heightThreshold: 148 },
  [DINCode.C]: { weightMin: 18, weightMax: 21, heightThreshold: 148 },
  [DINCode.D]: { weightMin: 22, weightMax: 25, heightThreshold: 148 },
  [DINCode.E]: { weightMin: 26, weightMax: 30, heightThreshold: 148 },
  [DINCode.F]: { weightMin: 31, weightMax: 35, heightThreshold: 157 },
  [DINCode.G]: { weightMin: 36, weightMax: 41, heightThreshold: 166 },
  [DINCode.H]: { weightMin: 42, weightMax: 48, heightThreshold: 174 },
  [DINCode.I]: { weightMin: 49, weightMax: 57, heightThreshold: 182 },
  [DINCode.J]: { weightMin: 58, weightMax: 66, heightThreshold: 189 },
  [DINCode.K]: { weightMin: 67, weightMax: 78, heightThreshold: 196 },
  [DINCode.L]: { weightMin: 79, weightMax: 94, heightThreshold: 196 },
  [DINCode.M]: { weightMin: 95, weightMax: 200, heightThreshold: Number.MAX_SAFE_INTEGER },
};

export const DIN_TABLE: DINTableEntry[] = [
  // Code A (10-13kg)
  { code: DINCode.A, bslRange: BSLRange.upTo250,      typeI: 0.75, typeII: 0.75, typeIII: 1.00 },
  { code: DINCode.A, bslRange: BSLRange.from251to270, typeI: 0.75, typeII: 0.75, typeIII: 1.00 },
  { code: DINCode.A, bslRange: BSLRange.from271to290, typeI: 0.75, typeII: 1.00, typeIII: 1.25 },
  { code: DINCode.A, bslRange: BSLRange.from291to310, typeI: 0.75, typeII: 1.00, typeIII: 1.25 },
  { code: DINCode.A, bslRange: BSLRange.from311to330, typeI: 0.75, typeII: 1.00, typeIII: 1.50 },
  { code: DINCode.A, bslRange: BSLRange.over330,      typeI: 0.75, typeII: 1.00, typeIII: 1.50 },

  // Code B (14-17kg)
  { code: DINCode.B, bslRange: BSLRange.upTo250,      typeI: 1.00, typeII: 1.25, typeIII: 1.50 },
  { code: DINCode.B, bslRange: BSLRange.from251to270, typeI: 1.00, typeII: 1.25, typeIII: 1.50 },
  { code: DINCode.B, bslRange: BSLRange.from271to290, typeI: 1.00, typeII: 1.50, typeIII: 1.75 },
  { code: DINCode.B, bslRange: BSLRange.from291to310, typeI: 1.00, typeII: 1.50, typeIII: 1.75 },
  { code: DINCode.B, bslRange: BSLRange.from311to330, typeI: 1.25, typeII: 1.50, typeIII: 2.00 },
  { code: DINCode.B, bslRange: BSLRange.over330,      typeI: 1.25, typeII: 1.50, typeIII: 2.00 },

  // Code C (18-21kg)
  { code: DINCode.C, bslRange: BSLRange.upTo250,      typeI: 1.50, typeII: 1.75, typeIII: 2.25 },
  { code: DINCode.C, bslRange: BSLRange.from251to270, typeI: 1.50, typeII: 1.75, typeIII: 2.25 },
  { code: DINCode.C, bslRange: BSLRange.from271to290, typeI: 1.50, typeII: 2.00, typeIII: 2.50 },
  { code: DINCode.C, bslRange: BSLRange.from291to310, typeI: 1.50, typeII: 2.00, typeIII: 2.50 },
  { code: DINCode.C, bslRange: BSLRange.from311to330, typeI: 1.75, typeII: 2.25, typeIII: 3.00 },
  { code: DINCode.C, bslRange: BSLRange.over330,      typeI: 1.75, typeII: 2.25, typeIII: 3.00 },

  // Code D (22-25kg)
  { code: DINCode.D, bslRange: BSLRange.upTo250,      typeI: 2.00, typeII: 2.50, typeIII: 3.00 },
  { code: DINCode.D, bslRange: BSLRange.from251to270, typeI: 2.00, typeII: 2.50, typeIII: 3.00 },
  { code: DINCode.D, bslRange: BSLRange.from271to290, typeI: 2.00, typeII: 2.75, typeIII: 3.50 },
  { code: DINCode.D, bslRange: BSLRange.from291to310, typeI: 2.00, typeII: 2.75, typeIII: 3.50 },
  { code: DINCode.D, bslRange: BSLRange.from311to330, typeI: 2.25, typeII: 3.00, typeIII: 3.50 },
  { code: DINCode.D, bslRange: BSLRange.over330,      typeI: 2.25, typeII: 3.00, typeIII: 3.50 },

  // Code E (26-30kg)
  { code: DINCode.E, bslRange: BSLRange.upTo250,      typeI: 2.50, typeII: 3.00, typeIII: 3.50 },
  { code: DINCode.E, bslRange: BSLRange.from251to270, typeI: 2.50, typeII: 3.00, typeIII: 3.50 },
  { code: DINCode.E, bslRange: BSLRange.from271to290, typeI: 2.75, typeII: 3.50, typeIII: 4.00 },
  { code: DINCode.E, bslRange: BSLRange.from291to310, typeI: 2.75, typeII: 3.50, typeIII: 4.00 },
  { code: DINCode.E, bslRange: BSLRange.from311to330, typeI: 3.00, typeII: 3.50, typeIII: 4.50 },
  { code: DINCode.E, bslRange: BSLRange.over330,      typeI: 3.00, typeII: 3.50, typeIII: 4.50 },

  // Code F (31-35kg)
  { code: DINCode.F, bslRange: BSLRange.upTo250,      typeI: 3.00, typeII: 3.50, typeIII: 4.50 },
  { code: DINCode.F, bslRange: BSLRange.from251to270, typeI: 3.00, typeII: 3.50, typeIII: 4.50 },
  { code: DINCode.F, bslRange: BSLRange.from271to290, typeI: 3.50, typeII: 4.00, typeIII: 5.00 },
  { code: DINCode.F, bslRange: BSLRange.from291to310, typeI: 3.50, typeII: 4.00, typeIII: 5.00 },
  { code: DINCode.F, bslRange: BSLRange.from311to330, typeI: 3.50, typeII: 4.50, typeIII: 5.50 },
  { code: DINCode.F, bslRange: BSLRange.over330,      typeI: 3.50, typeII: 4.50, typeIII: 5.50 },

  // Code G (36-41kg)
  { code: DINCode.G, bslRange: BSLRange.upTo250,      typeI: 3.50, typeII: 4.50, typeIII: 5.50 },
  { code: DINCode.G, bslRange: BSLRange.from251to270, typeI: 3.50, typeII: 4.50, typeIII: 5.50 },
  { code: DINCode.G, bslRange: BSLRange.from271to290, typeI: 4.00, typeII: 5.00, typeIII: 6.00 },
  { code: DINCode.G, bslRange: BSLRange.from291to310, typeI: 4.00, typeII: 5.00, typeIII: 6.00 },
  { code: DINCode.G, bslRange: BSLRange.from311to330, typeI: 4.50, typeII: 5.50, typeIII: 6.50 },
  { code: DINCode.G, bslRange: BSLRange.over330,      typeI: 4.50, typeII: 5.50, typeIII: 6.50 },

  // Code H (42-48kg)
  { code: DINCode.H, bslRange: BSLRange.upTo250,      typeI: 4.50, typeII: 5.50, typeIII: 6.50 },
  { code: DINCode.H, bslRange: BSLRange.from251to270, typeI: 4.50, typeII: 5.50, typeIII: 6.50 },
  { code: DINCode.H, bslRange: BSLRange.from271to290, typeI: 5.00, typeII: 6.00, typeIII: 7.50 },
  { code: DINCode.H, bslRange: BSLRange.from291to310, typeI: 5.00, typeII: 6.00, typeIII: 7.50 },
  { code: DINCode.H, bslRange: BSLRange.from311to330, typeI: 5.50, typeII: 6.50, typeIII: 8.00 },
  { code: DINCode.H, bslRange: BSLRange.over330,      typeI: 5.50, typeII: 6.50, typeIII: 8.00 },

  // Code I (49-57kg)
  { code: DINCode.I, bslRange: BSLRange.upTo250,      typeI: 5.50, typeII: 6.50, typeIII: 8.00 },
  { code: DINCode.I, bslRange: BSLRange.from251to270, typeI: 5.50, typeII: 6.50, typeIII: 8.00 },
  { code: DINCode.I, bslRange: BSLRange.from271to290, typeI: 6.00, typeII: 7.50, typeIII: 9.00 },
  { code: DINCode.I, bslRange: BSLRange.from291to310, typeI: 6.00, typeII: 7.50, typeIII: 9.00 },
  { code: DINCode.I, bslRange: BSLRange.from311to330, typeI: 6.50, typeII: 8.00, typeIII: 9.50 },
  { code: DINCode.I, bslRange: BSLRange.over330,      typeI: 6.50, typeII: 8.00, typeIII: 9.50 },

  // Code J (58-66kg)
  { code: DINCode.J, bslRange: BSLRange.upTo250,      typeI: 6.50, typeII: 8.00, typeIII: 9.50 },
  { code: DINCode.J, bslRange: BSLRange.from251to270, typeI: 6.50, typeII: 8.00, typeIII: 9.50 },
  { code: DINCode.J, bslRange: BSLRange.from271to290, typeI: 7.50, typeII: 9.00, typeIII: 10.50 },
  { code: DINCode.J, bslRange: BSLRange.from291to310, typeI: 7.50, typeII: 9.00, typeIII: 10.50 },
  { code: DINCode.J, bslRange: BSLRange.from311to330, typeI: 8.00, typeII: 9.50, typeIII: 11.00 },
  { code: DINCode.J, bslRange: BSLRange.over330,      typeI: 8.00, typeII: 9.50, typeIII: 11.00 },

  // Code K (67-78kg)
  { code: DINCode.K, bslRange: BSLRange.upTo250,      typeI: 8.00,  typeII: 9.50,  typeIII: 11.00 },
  { code: DINCode.K, bslRange: BSLRange.from251to270, typeI: 8.00,  typeII: 9.50,  typeIII: 11.00 },
  { code: DINCode.K, bslRange: BSLRange.from271to290, typeI: 8.50,  typeII: 10.00, typeIII: 12.00 },
  { code: DINCode.K, bslRange: BSLRange.from291to310, typeI: 8.50,  typeII: 10.00, typeIII: 12.00 },
  { code: DINCode.K, bslRange: BSLRange.from311to330, typeI: 9.00,  typeII: 11.00, typeIII: 12.00 },
  { code: DINCode.K, bslRange: BSLRange.over330,      typeI: 9.00,  typeII: 11.00, typeIII: 12.00 },

  // Code L (79-94kg)
  { code: DINCode.L, bslRange: BSLRange.upTo250,      typeI: 9.00,  typeII: 11.00, typeIII: 12.00 },
  { code: DINCode.L, bslRange: BSLRange.from251to270, typeI: 9.00,  typeII: 11.00, typeIII: 12.00 },
  { code: DINCode.L, bslRange: BSLRange.from271to290, typeI: 9.50,  typeII: 11.00, typeIII: 12.00 },
  { code: DINCode.L, bslRange: BSLRange.from291to310, typeI: 9.50,  typeII: 11.00, typeIII: 12.00 },
  { code: DINCode.L, bslRange: BSLRange.from311to330, typeI: 10.00, typeII: 12.00, typeIII: 12.00 },
  { code: DINCode.L, bslRange: BSLRange.over330,      typeI: 10.00, typeII: 12.00, typeIII: 12.00 },

  // Code M (95+kg)
  { code: DINCode.M, bslRange: BSLRange.upTo250,      typeI: 10.00, typeII: 12.00, typeIII: 12.00 },
  { code: DINCode.M, bslRange: BSLRange.from251to270, typeI: 10.00, typeII: 12.00, typeIII: 12.00 },
  { code: DINCode.M, bslRange: BSLRange.from271to290, typeI: 11.00, typeII: 12.00, typeIII: 12.00 },
  { code: DINCode.M, bslRange: BSLRange.from291to310, typeI: 11.00, typeII: 12.00, typeIII: 12.00 },
  { code: DINCode.M, bslRange: BSLRange.from311to330, typeI: 12.00, typeII: 12.00, typeIII: 12.00 },
  { code: DINCode.M, bslRange: BSLRange.over330,      typeI: 12.00, typeII: 12.00, typeIII: 12.00 },
];

export function bslRangeFrom(bslMm: number): BSLRange {
  if (bslMm <= 250) return BSLRange.upTo250;
  if (bslMm <= 270) return BSLRange.from251to270;
  if (bslMm <= 290) return BSLRange.from271to290;
  if (bslMm <= 310) return BSLRange.from291to310;
  if (bslMm <= 330) return BSLRange.from311to330;
  return BSLRange.over330;
}

export function dinTableLookup(code: DINCode, bslMm: number, ability: AbilityLevel): number {
  const range = bslRangeFrom(bslMm);
  const entry = DIN_TABLE.find((e) => e.code === code && e.bslRange === range);
  if (!entry) return 1.0;
  switch (ability) {
    case AbilityLevel.Beginner:
      return entry.typeI;
    case AbilityLevel.Intermediate:
      return entry.typeII;
    case AbilityLevel.Advanced:
      return entry.typeIII;
  }
}
